dev/PipsControl/PipsControl.cpp

```diff
@@ -0,0 +1,335 @@
+        {
+            winrt::get_self<PipsControlAutomationPeer>(peer)->RaiseSelectionChanged(m_lastSelectedPageIndex, SelectedPageIndex());
+        }
+        UpdateNavigationButtonVisualStates();
+        UpdateVerticalPips(NumberOfPages(), SelectedPageIndex());
+        RaiseSelectedIndexChanged();
+    }
+}
+
+void PipsControl::RaiseSelectedIndexChanged()
+{
+    const auto args = winrt::make_self<PipsControlSelectedIndexChangedEventArgs>(m_lastSelectedPageIndex, SelectedPageIndex());
+    m_selectedIndexChangedEventSource(*this, *args);
+}
+
+void PipsControl::MovePipIdentifierToElement(int index) {
+
+    if (NumberOfPages() != 0) {
+        if (const auto repeater = m_verticalPipsRepeater.get())
+        {
+            if (const auto element = repeater.TryGetElement(m_lastSelectedPageIndex).try_as<winrt::Button>()) {
+                winrt::VisualStateManager::GoToState(element, L"Unselected", true);
+            }
+            if (const auto element = repeater.GetOrCreateElement(index).try_as<winrt::Button>()) {
+                element.UpdateLayout();
+                winrt::VisualStateManager::GoToState(element, L"Selected", true);
+                ScrollToCenterOfViewport(element);
+            }
+        }
+    }
+}
+
+void PipsControl::setVerticalPipsSVMaxSize() {
+    auto pipHeight = unbox_value<double>(ResourceAccessor::ResourceLookup(*this, box_value(c_PipsControlButtonHeightPropertyName)));
+    auto numberOfPages = NumberOfPages() < 0 ? MaxDisplayedPages() : std::min(NumberOfPages(), MaxDisplayedPages());
+    auto scrollViewerHeight = pipHeight * numberOfPages;
+    m_verticalPipsScrollViewer.get().MaxHeight(scrollViewerHeight);
+}
+
+
+
+void PipsControl::UpdateVerticalPips(const int numberOfPages, const int maxDisplayedPages) {
+
+    auto pipsListSize = static_cast<int>(m_verticalPipsElements.Size());
+    auto const selectedIndex = SelectedPageIndex();
+
+    if (numberOfPages < pipsListSize) {
+        m_verticalPipsElements.Clear();
+        pipsListSize = 0;
+    }
+
+    auto const endIndex = std::min(numberOfPages, selectedIndex + maxDisplayedPages);
+    for (int i = pipsListSize; i < endIndex; i++) {
+        m_verticalPipsElements.Append(box_value(i + 1));
+    }
+ 
+    if (maxDisplayedPages != m_lastMaxDisplayedPages) {
+        setVerticalPipsSVMaxSize();
+    }
+    MovePipIdentifierToElement(SelectedPageIndex());
+
+}
+
+void PipsControl::OnRootGridKeyDown(const winrt::IInspectable & sender, const winrt::KeyRoutedEventArgs & args) {
+    if (args.Key() == winrt::VirtualKey::Left || args.Key() == winrt::VirtualKey::Up)
+    {
+        winrt::FocusManager::TryMoveFocus(winrt::FocusNavigationDirection::Up);
+    }
+    else if (args.Key() == winrt::VirtualKey::Right || args.Key() == winrt::VirtualKey::Down)
+    {
+        winrt::FocusManager::TryMoveFocus(winrt::FocusNavigationDirection::Down);
+    }
```

Comment:
Would just setting the XYFocus properties on the repeater achieve this ?

Reply:
I think so, I also think this would probably make for a bad gamepad experience as written, it doesn't look like there is a way to move left or right from the vertical pips without navigating to the end of the pips list.

---
In reply to: [520925092](https://github.com/microsoft/microsoft-ui-xaml/pull/3592#discussion_r520925092) [](ancestors = 520925092)

Reply:
Ah, that's a good point for the Gamepad experiences. Can we have a different interaction model for keyboarding versus Gamepad spatial nav? For keyboard input, especially when the user is using a screen reader, it seems like a good idea for left/up and right/down to have the same behavior. For Gamepad, I could see us limiting it to only up/down for vertical orientation and left/right for horizontal orientation.